accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+
+```dart
+// Emulate the inline class using a class.
+
+class TinyJson {
+  // The representation is assumed to be a nested list of numbers.
+  final Object it;
+
+  TinyJson(this.it);
+
+  Iterable<num> get leaves sync* {
+    var localIt = it; // To get promotion.
+    if (localIt is num) {
+      yield localIt;
+    } else if (localIt is List<dynamic>) {
+      for (var element in localIt) {
+        yield* TinyJson(element).leaves;
+      }
+    } else {
+      throw "Unexpected object encountered in TinyJson value";
+    }
+  }
+}
+
+void main() {
+  var tiny = TinyJson(<dynamic>[<dynamic>[1, 2], 3, <dynamic>[]]);
+  print(tiny.leaves);
+  tiny.add("Hello!"); // Error.
+}
+```
+
+This is similar to the inline type in that it enforces the use of
+specific operations only (here we have just one: `leaves`), and it
+makes it an error to use instance methods of the representation (e.g.,
+`add`).
+
+Creation of wrapper objects consumes space and time. In the case where
+we wish to work on an entire data structure, we'd need to wrap each
+object as we navigate the data structure. For instance, we need to
+create a wrapper `TinyJson(element)` in order to invoke `leaves`
+recursively.
+
+In contrast, the inline class is zero-cost, in the sense that it does
+_not_ use a wrapper object, it enforces the desired discipline
+statically. In the inline class, the invocation of `TinyJson(element)`
+in the body of `leaves` can be eliminated entirely by inlining.
+
+Inline classes are static in nature, like extension members: an inline
+class declaration may declare some type parameters. The type
+parameters will be bound to types which are determined by the static
+type of the receiver. Similarly, members of an inline type are resolved
+statically, i.e., if `tiny.leaves` is an invocation of an inline getter
+`leaves`, then the declaration named `leaves` whose body is executed
+is determined at compile-time. There is no support for late binding of
+an inline member, and hence there is no notion of overriding. In return
+for this lack of expressive power, we get improved performance.
+
+
+## Syntax
+
+A rule for `<inlineClassDeclaration>` is added to the grammar, along
+with some rules for elements used in inline class declarations:
+
+```ebnf
+<inlineClassDeclaration> ::=
+  'inline' 'class' <typeIdentifier> <typeParameters>? <interfaces>?
+  '{'
+    (<metadata> <inlineMemberDeclaration>)*
+  '}'
+
+<inlineMemberDeclaration> ::= <classMemberDefinition>
+```
+
+*The token `inline` is not made a built-in identifier: the reserved
+word `class` that occurs right after `inline` serves to disambiguate
+the inline class declaration with a fixed lookahead.*
+
+A few errors can be detected immediately from the syntax:
+
+A compile-time error occurs unless the inline class declares exactly
+one instance variable. Let `v` be the name of said instance
+variable. A compile-time error occurs unless the declaration of `v`
+has the modifier `final` or the modifier `late`.
```

Comment:
Do you mean ", or if it has the modifier `late`"? 
Here the "or" is an alternative to having the modifier `final`, so `late int i` is valid.

Reply:
Consider rewriting to something like "The declaration of `v` must have the modifier `final` and must not have the modifier `late`; otherwise a compile-time error occurs."

Reply:
This. I dislike the tabloid headline style of "Disaster Will Happen ... unless very likely caveat".

Say what must be true first. Then say what happens if it isn't. If you only read the first part, your code will still be fine.

Reply:
> This. I dislike...

Ah, I did not respond to this comment. (It was not shown when I responded to the review comments yesterday).

We do have an impedance mismatch here, though.

I recognize that it may read quite well to say "Do this and your program is fine" and then something like "Otherwise, an error occurs".

However, that style is somewhat counterproductive because it requires a specification of "good" programs, and we otherwise never do that (at least in the language specification). We specify the conditions that give rise to a compile-time error, and we specify the properties and behavior of specific shapes of program elements like expressions, statements, and declarations (and this behavior may include some dynamic errors). It is always understood that the latter is only applicable to programs with no compile-time errors.

The point is that the set of programs with no compile-time errors is determined by taking all syntactically correct programs and removing some of them (because they have non-syntax, compile-time errors). The remaining set is complex, and it consists of contributions from many different parts of the specification (e.g., for expressions we have many different sections, and each of them contribute a set of correct expressions with a specific syntactic form).

One thing which would be possible, and which wouldn't trigger this "good programs" problem, would be to always specify the error conditions positively (for instance, we wouldn't use 'unless').

Did that.